pkg/ledger/executor.go

```diff
@@ -98,65 +103,85 @@ func (l *Ledger) execute(ctx context.Context, script core.Script) (*core.Transac
 		case vm.EXIT_FAIL_INVALID:
 			return nil, errors.New("internal error: compiled script was invalid")
 		case vm.EXIT_FAIL_INSUFFICIENT_FUNDS:
-			// TODO: If the machine can provide the asset which is failing, we should be able to use InsufficientFundError{} instead of error code
-			return nil, NewScriptError(ScriptErrorInsufficientFund, "account had insufficient funds")
+			// TODO: If the machine can provide the asset which is failing
+			// we should be able to use InsufficientFundError{} instead of error code
+			return nil, NewScriptError(ScriptErrorInsufficientFund,
+				"account had insufficient funds")
 		default:
 			return nil, errors.New("script execution failed")
 		}
 	}
 
-	metadata := m.GetTxMetaJson()
-	for k, v := range metadata {
+	txMeta := m.GetTxMetaJSON()
+	for k, v := range txMeta {
 		asMapAny := make(map[string]any)
-		err := json.Unmarshal(v.([]byte), &asMapAny)
-		if err != nil {
+		if err := json.Unmarshal(v.([]byte), &asMapAny); err != nil {
 			panic(err)
 		}
-		metadata[k] = asMapAny
+		txMeta[k] = asMapAny
 	}
 	for k, v := range script.Metadata {
-		_, ok := metadata[k]
+		_, ok := txMeta[k]
 		if ok {
-			return nil, NewScriptError(ScriptErrorMetadataOverride, "cannot override metadata from script")
+			return nil, NewScriptError(ScriptErrorMetadataOverride,
+				"cannot override metadata from script")
 		}
-		metadata[k] = v
-	}
-
-	t := &core.TransactionData{
-		Postings:  m.Postings,
-		Metadata:  metadata,
-		Reference: script.Reference,
+		txMeta[k] = v
 	}
 
-	return t, nil
-}
-
-func (l *Ledger) Execute(ctx context.Context, script core.Script) (*core.ExpandedTransaction, error) {
-
-	t, err := l.execute(ctx, script)
-	if err != nil {
-		return nil, err
+	accMeta := core.Metadata{}
+	for account, meta := range m.GetAccountsMetaJSON() {
+		meta := meta.(map[string][]byte)
+		for k, v := range meta {
+			asMapAny := make(map[string]any)
+			if err := json.Unmarshal(v, &asMapAny); err != nil {
+				return nil, errors.Wrap(err, "json.Unmarshal")
+			}
+			if _, ok := accMeta["set_account_meta"]; !ok {
+				accMeta["set_account_meta"] = map[string]any{}
+			}
+			if account[0] == '@' {
+				account = account[1:]
+			}
+			if _, ok := accMeta["set_account_meta"].(map[string]any)[account]; !ok {
+				accMeta["set_account_meta"].(map[string]any)[account] = map[string]any{}
+			}
+			accMeta["set_account_meta"].(map[string]any)[account].(map[string]any)[k] = asMapAny
+		}
 	}
 
-	res, err := l.Commit(ctx, []core.TransactionData{*t})
-	if err != nil {
-		return nil, err
+	postings := make([]core.Posting, len(m.Postings))
+	for i, p := range m.Postings {
+		amt := core.MonetaryInt(*p.Amount)
+		postings[i] = core.Posting{
+			Source:      p.Source,
+			Destination: p.Destination,
+			Amount:      &amt,
+			Asset:       p.Asset,
+		}
 	}
 
-	return &res.GeneratedTransactions[0], nil
+	return &core.TransactionData{
+		Postings:  postings,
+		Metadata:  core.Metadata(txMeta).Merge(accMeta),
```

Comment:
Mmm, account metadata should be set on accounts and not on transaction no?

Reply:
I decided, in addition to set metadata on accounts (which I am doing in the `ledger.Commit` method), I am also setting this additional operation in the transaction metadata, under a `set_account_meta` key, to keep a trace in the transaction. Don't you think it could be useful?

Reply:
I guess it is to save on transactions metadata, metadata we want to add for accounts, as accounts metadata are inserted later on accounts.
Apart if we have a need for this, i don't think we should support unneeded features.
It is more simple to add a feature than removing it ;)